### ISSUE TITLE
clojure-lsp silently stops working in calva after a while

### ISSUE THREAD

**Issue by calherries** [BetterThanTomorrow/calva#1782]
Note: duplicated issue in calva repo: https://github.com/BetterThanTomorrow/calva/issues/1782

**Describe the bug**
After about 10 minutes after starting clojure-lsp and using the editor, it’s functionality in the editor stops. I can solve it temporarily by restarting clojure-lsp, but it soon fails again.

The issue persists after disconnecting with the REPL.

Things that stop working:
- Any new syntax errors I create in the editor aren’t underlined with linting errors (existing ones before it stops working persist)
- Any new errors I create aren’t showing in the “Problems” tab (but existing ones persist)
- Go to definition
- All code navigation and refactoring commands
- Things that are still working:

The clojure-lsp log file still logs handlers, publishing diagnostics, etc. If I intentionally create a syntax error I can see the logs update with a correctly updated “Publishing X diagnostics” message
```
2022-06-24T09:45:14.936Z DEBUG [clojure-lsp.server:82] - Publishing 11 diagnostics for file:///Users/callumherries/meta/metabase/src/metabase/query_processor/card.clj
```
VS code still shows clojure-lsp as working, and when I click the button, it shows the relevant quick actions.

**To Reproduce**
It happens reliably on my machine after jumping between files, definitions, etc., so I don’t know the specific cause. I haven’t tried with a minimal example but it happens while I’ve been working on https://github.com/metabase/metabase, which is open source.

<details>
 <summary><b>Log - client <-> server</b></summary>
<pre>
[Trace - 10:23:59 PM] Sending notification '$/cancelRequest'.
Params: {
    "id": 808
}

Jun 24, 2022 10:23:59 PM org.eclipse.lsp4j.jsonrpc.RemoteEndpoint handleCancellation
WARNING: Unmatched cancel notification for request id 808
[Trace - 10:23:59 PM] Sending request 'textDocument/codeAction - (809)'.
Params: {
    "textDocument": {
        "uri": "file:///Users/callumherries/meta/metabase/src/metabase/query_processor/card.clj"
    },
    "range": {
        "start": {
            "line": 249,
            "character": 23
        },
        "end": {
            "line": 249,
            "character": 23
        }
    },
    "context": {
        "diagnostics": []
    }
}

[Trace - 10:23:59 PM] Sending notification '$/cancelRequest'.
Params: {
    "id": 807
}

Jun 24, 2022 10:23:59 PM org.eclipse.lsp4j.jsonrpc.RemoteEndpoint handleCancellation
WARNING: Unmatched cancel notification for request id 807
[Trace - 10:24:16 PM] Sending request 'textDocument/codeAction - (810)'.
Params: {
    "textDocument": {
        "uri": "file:///Users/callumherries/meta/metabase/src/metabase/query_processor/card.clj"
    },
    "range": {
        "start": {
            "line": 249,
            "character": 19
        },
        "end": {
            "line": 249,
            "character": 19
        }
    },
    "context": {
        "diagnostics": []
    }
}

[Trace - 10:24:16 PM] Sending notification '$/cancelRequest'.
Params: {
    "id": 809
}

Jun 24, 2022 10:24:16 PM org.eclipse.lsp4j.jsonrpc.RemoteEndpoint handleCancellation
WARNING: Unmatched cancel notification for request id 809
[Trace - 10:24:16 PM] Sending notification '$/cancelRequest'.
Params: {
    "id": 805
}

[Trace - 10:24:16 PM] Sending request 'textDocument/documentHighlight - (811)'.
Params: {
    "textDocument": {
        "uri": "file:///Users/callumherries/meta/metabase/src/metabase/query_processor/card.clj"
    },
    "position": {
        "line": 248,
        "character": 19
    }
}

Jun 24, 2022 10:24:16 PM org.eclipse.lsp4j.jsonrpc.RemoteEndpoint handleCancellation
WARNING: Unmatched cancel notification for request id 805
[Trace - 10:24:16 PM] Sending request 'textDocument/codeAction - (812)'.
Params: {
    "textDocument": {
        "uri": "file:///Users/callumherries/meta/metabase/src/metabase/query_processor/card.clj"
    },
    "range": {
        "start": {
            "line": 248,
            "character": 19
        },
        "end": {
            "line": 248,
            "character": 19
        }
    },
    "context": {
        "diagnostics": []
    }
}

[Trace - 10:24:16 PM] Sending notification '$/cancelRequest'.
Params: {
    "id": 810
}

Jun 24, 2022 10:24:16 PM org.eclipse.lsp4j.jsonrpc.RemoteEndpoint handleCancellation
WARNING: Unmatched cancel notification for request id 810
[Trace - 10:24:16 PM] Sending request 'textDocument/definition - (813)'.
Params: {
    "textDocument": {
        "uri": "file:///Users/callumherries/meta/metabase/src/metabase/query_processor/card.clj"
    },
    "position": {
        "line": 248,
        "character": 19
    }
}

</pre>
</details>

**User details (please complete the following information):**
 - OS: MacOS
 - Editor VSCode (Calva) (version from 13/6/2022, 17:40:41)
 - Version: 2022.06.22-14.09.50

**Comment by bpringe** [comment]
I think the same thing has been happening to me today, which seems to be a new issue. Are those logs from a time period in which clojure-lsp functionality was not working? I see a lot of cancel requests. I don't know if that's related or what those are, really.

@ericdallo Have you heard any reports of similar behavior? Does anything in the logs above stand out?

@calherries Can you post your server logs next time you experience the issue? (https://calva.io/clojure-lsp/#opening-the-server-log-file)

**Comment by bpringe** [comment]
I just had this happen again. The command for opening the log file won't work when this happens either. I noticed we get no response from the `clojure/serverInfo/raw` request.

```
Jun 24, 2022 3:37:49 PM org.eclipse.lsp4j.jsonrpc.RemoteEndpoint handleCancellation
WARNING: Unmatched cancel notification for request id 556
[Trace - 3:38:12 PM] Sending request 'clojure/serverInfo/raw - (558)'.
[Trace - 3:39:25 PM] Sending notification '$/cancelRequest'.
Params: {
    "id": 557
}


Jun 24, 2022 3:39:25 PM org.eclipse.lsp4j.jsonrpc.RemoteEndpoint handleCancellation
WARNING: Unmatched cancel notification for request id 557
```

**Comment by bpringe** [comment]
My info:

OS: MacOS
Editor: VSCode version 1.68.1
Calva Version: 2.0.286
Clojure-lsp Version: 2022.06.22-14.09.50

**Comment by ericdallo** [comment]
@bpringe we are analyzing this issue [here](https://clojurians.slack.com/archives/CPABC1H61/p1656093769590509?thread_ts=1656093769.590509&cid=CPABC1H61).
Would you mind posting in that thread as well ?

**Comment by bpringe** [comment]
For anyone reading this who has this problem, a workaround for now is to [make Calva use an earlier version of clojure-lsp](https://calva.io/clojure-lsp/#changing-the-version-of-clojure-lsp-used-by-calva), such as `2022.05.31-17.35.50`.

**Comment by davinomjr** [comment]
Happening to me as well, log from client is similar to what @bpringe posted. I couldnt find a pattern as it happens even without REPL (soon after a restart on the editor). 

Btw the workaround of Calva using an earlier version of `clojure-lsp` fixed it for now.

**Comment by sebhs** [comment]
For anyone who tried using an earlier version of clojure-lsp and it didn't work, you actually also need to download the old version of clojure-lsp after you update the settings


- stop clojure-lsp
- download clojure-lsp
- start clojure-lsp
(Using the status bar clojure-lsp button.)

**Comment by bpringe** [comment]
A reload of the window will basically do the above steps for you, just FYI. (There's a command to reload the window.)

**Comment by PEZ** [comment]
I'd like to promote the manual stop->download->start way of doing it instead of reloading the window, which does a lot more things than change the clojure-lsp version used.

In any case, the important thing to take away from @sebhs's comment is that just updating the version in settings isn't enough. We probably should update the setting description.

**Comment by PEZ** [comment]
I now note that we do inform that it takes a reload to make the setting take effect.

![image](https://user-images.githubusercontent.com/30010/176379075-d2f6f925-12d6-4cd7-953c-5a2a209d515b.png)

So, that's where we should be informing about the proper way to do it.

**Comment by bpringe** [comment]
This was fixed in [2022.06.29-19.32.13](https://github.com/clojure-lsp/clojure-lsp/releases/tag/2022.06.29-19.32.13).

**Comment by PEZ** [comment]
This is now fixed with version [2022.06.29-19.32.13](https://github.com/clojure-lsp/clojure-lsp/releases/tag/2022.06.29-19.32.13) of clojure-lsp.

Thanks @mainej and @ericdallo for being so quick to give this attention and find and implement the solution! 🙏 ❤️
